Avoid dynamic allocations for single-point Variables data

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Fully suppress colors with --color=no

wthrowe committed:
Store 1-point Variables data in static allocation

This is an optimization to avoid allocations when solving small
systems of ODEs.

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -212,8 +212,10 @@ class DataBox<tmpl::list<Tags...>> : private detail::Item<Tags>... {
   DataBox() = default;
-  DataBox(DataBox&& rhs) = default;
-  DataBox& operator=(DataBox&& rhs) {
+  constexpr DataBox(DataBox&& rhs) : detail::Item<Tags>(std::move(rhs))... {
```
can we add a test that would have failed without this change?

wthrowe replied:
I've modified Test_DataBox to check this.  Didn't make it a fixup because these are the only changes to that file in this PR.

nilsdeppe reviewed (dismissed):
LGTM, no additional requests. I don't need to re-approve for it to get merged.

wthrowe committed:
Reset subitems after moving a DataBox

Subitems can store pointers into their parent items, which may be
invalidated by a move.

kidder commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -1865,39 +1865,27 @@ namespace {
-      p | *data_;
-    }
-  }
+  void pup(PUP::er& p) { p | data_; }
```
should  `ptr_` be reset if `p.isUnpacking()`?

wthrowe replied:
`&data_` hasn't changed.
